### ISSUE TITLE
Missing some events on CNCFtest

### ISSUE THREAD

**Issue by jberkus** [cncf/devstats.archive#111]
Lukasz:

I found a case of a missed event causing us to miss a change in issue status.  Take a look at:

https://github.com/kubernetes/kubernetes/issues/59001
This is issue ID 292617119 in the database.

This issue was demoted in priority 3 days ago, when yujuhong made a comment.  Thing is, in the database, the last recorded comment is @tpepper's from 9 days ago, which by my count means we're missing 3 events.  Because there was no event, the check on labels didn't fire.

**Comment by lukaszgryglicki** [comment]
GHAPI2DB should catch this. I'll investigate on Monday.

**Comment by lukaszgryglicki** [comment]
Investigating this.

**Comment by lukaszgryglicki** [comment]
I've regenerated data for 2018-05-22 - 2018-05-26 on a temp database and saved all JSON's that I've received from GHA.
None of them contain any event related to [this](https://github.com/kubernetes/kubernetes/issues/59001#issuecomment-391073109) comment.
So it seems like a bug in GitHub archives.
I'll report this on GHA site.
For this issue I'll run ghaap2db manually specifying this issue (ghapi2db also has special manual mode that allows to process list of issues provided from command line).

**Comment by lukaszgryglicki** [comment]
Manually updated issue via `GHA2DB_ONLY_ISSUES="292617119" GHA2DB_DEBUG=1 ./ghapi2db`.
Issue state is now correct in DevStats database.
Reporting bug on GHA.

**Comment by lukaszgryglicki** [comment]
Issue on github archives created: https://github.com/igrigorik/gharchive.org/issues/187

**Comment by lukaszgryglicki** [comment]
I'm adding blocked label. I've also asked if this is a very rare case, or maybe GHA just miss some % of events?

**Comment by jberkus** [comment]
Yeah, the main thing for us at this stage is to figure out how often this happens.

**Comment by lukaszgryglicki** [comment]
Exactly +1

**Comment by lukaszgryglicki** [comment]
@jberkus take a look at [this](https://github.com/igrigorik/gharchive.org/issues/171)

**Comment by jberkus** [comment]
So, where does this leave us?  It kinda seems like we've just discovered that our primary way to get data out of github isn't reliable.

**Comment by lukaszgryglicki** [comment]
Maybe we can help to fix this?

**Comment by lukaszgryglicki** [comment]
@jberkus do we also miss PR merge events?
I'm quite sure we don't because k8s-bot comments when merging from submit-queue.
Asking because all Repos/Issues/Events API that I've tried are not returning PR state.
The only info I can get from those APIs is event type = "merged"
For `github.go` sources:
```
//     merged
//       The Actor merged into master a branch containing a commit mentioning the issue.
//       CommitID holds the SHA1 of the merge commit.
```
I was about to try to create some smart PR state objects (as mirror to the Issue) but I'm quite sure that we receive PR merge events.

Speaking about *merge* not *close*.
We can now detect close via API calls because PR structure in GitHub is quite strange.
PR is an issue (so it have entry in `gha_issues` with `is_pull_request = true`, but also additional state in `gha_pull_requests` which contains data specific only to PRs. And API is not returing that data to me.

**Comment by lukaszgryglicki** [comment]
@jberkus nevermind, after a lot of research I see that for issues that are PRs `IsPullRequest() == true` I would have to call GH PR API anyway - to make sure we have a correct state for a PR.

**Comment by lukaszgryglicki** [comment]
@dankohn @jberkus finally after another 13-14 hours struggling I think I may have something that will work for us.
I've installed cron on the test and will see results tomorrow - hopefully after 10 successful syncs.
It should handle all we need: changing state of
- Issue
- PR
Due to
- Labels, Milestone, state: Closed/Reopened, Renamed and much more
I'm using GH API to detect what changed recently, if changed issue is a PR I'm doing additional PR specific sync.

Everything here:
- https://github.com/cncf/devstats/blob/master/cmd/ghapi2db/ghapi2db.go
- https://github.com/cncf/devstats/blob/master/ghapi.go

For details see all my commits from today:
https://github.com/cncf/devstats/commits/master

**Comment by lukaszgryglicki** [comment]
It is even better than what we have before.
Before we were querying current issues state - that were modified recently. But we could miss the updated issue when no event was recorded - it wasn't treated as recently modified then.

Now we're querying GitHub events API to get recent events and then we're getting issues state from the past and for the PRs (issues which are PRs) we're additionally querying for most up to date state (there is no PRs events API, so we can only get current not past state, but this is for PRs/Issues not older than last hour).

But the biggest gain is that we will also have all events that GHA **cannot** register anyway. Those generated by changing labels, milestones etc without commenting.
This is the list of events that will be recorded using GitHub Issue Events API (which are not available for GHA):

- Issue: locked, unlocked, renamed, assigned, reassigned, unassigned, label changed (added, removed, changed), milestoned (milestone changed, added, removed).

So even when GHA is fixed - this will stay there and will be **very** useful.
Those new event types are just events of a different kind - they're not recorded by GHA.

**Comment by lukaszgryglicki** [comment]
Can we close this?

**Comment by jberkus** [comment]
Nope, we are still missing events.

From Github UI:

```
is:pr is:open milestone:v1.12 
6 Open 16 Closed
```

From database:

```
gha=> select state, count(*) from issues where is_pull_request and milestone='v1.12' group by state;
 state  | count 
--------+-------
 closed |     9
 open   |    15
```

So it's *still* the case that close/merge/change milestone events are not getting through to the database.

Three example PRs which are closed in GH, but not in Devstats:

* 56358
* 62692
* 64883

**Comment by jberkus** [comment]
FWIW, it does look like all PRs assigned to the milestone are being recorded.  It's just closing them or changing their milestone which isn't.

**Comment by lukaszgryglicki** [comment]
I'll take a look tomorrow at first glance:
- We may have up to hour lag before devstats get data for a given hour.
- Some events could happen before we started to use GH API to get this new events (but such event must have happened more than 16-18 days ago), so we're not in sync for them yet
- Your select refers to non-existing tables - this is some shortcut, right? Anyway I'll check those PRs state carefully tomorrow (they are k/k PRs?)

**Comment by jberkus** [comment]
The "issues" table is a materialized view in the devstats_team schema.

> We may have up to hour lag before devstats get data for a given hour.

None of these events occurred in the last hour (the one for the first PR was 20 days ago).

> Some events could happen before we started to use GH API to get this new events (but such event must have happened more than 16-18 days ago), so we're not in sync for them yet

Ah, yeah, if we're not in sync for those, that could explain it.  I'll look specifically for missing events in the last week.

Yes, they're k/k PRs.

**Comment by lukaszgryglicki** [comment]
I'll take a look today to confirm.
We also *may* loose some events on the test server when for example devstats is stopped because I'm doing some work.
But I've added a code that backups all those events coming from GH API every day with the database backup - so I can restore them from prod for example.
And in case we need to regenerate all data from scratch we won't have any of those events - so in this case I can regenerate database and then add all GH API events gathered so far from a special `tsv` backup.
See:
- https://github.com/cncf/devstats/blob/master/devel/backup_artificial.sh
- https://github.com/cncf/devstats/blob/master/devel/backup_artificial_all.sh
- https://github.com/cncf/devstats/blob/master/devel/restore_artificial.sh
- https://github.com/cncf/devstats/blob/master/devel/restore_artificial_all.sh

**Comment by lukaszgryglicki** [comment]
I'll probably take a look at this tomorrow, I have a lot of work with updating affiliations (including some unexpected small bug fixes and then retries). Sorry.

**Comment by lukaszgryglicki** [comment]
From GitHub UI: `is:pr is:open milestone:v1.12` -> `6 Open 16 Closed`.
From your MV (materialized view) after refresh: `select state, count(*) from current_state.issues where is_pull_request and milestone='v1.12' group by state` --> `Open 14 Closed 9`.

`select string_agg(number::text, ', ') from current_state.issues where is_pull_request and milestone='v1.12' and state = 'open/closed'`

Open in GH: 65660, 64662, 64283, 64005, 63947, 63437
Open in Devstats: 56358, 57932, 62723, 63437, 63625, 63947, 64005, 64283, 64662, 64831, 64837, 64883, 65601, 65660
Closed in GH: 65601, 65380, 65027, 65001, 64938, 64883, 64837, 64831, 64752, 64539, 64246, 63625, 63461, 61718, 57932, 56358
Closed in Devstats: 61718, 63461, 64246, 64539, 64752, 64938, 65001, 65027, 65380

Both open: 65660, 64662, 64283, 64005, 63947, 63437
Both closed: 65380, 65027, 65001, 64938, 64752, 64539, 64246, 63461, 61718

Open in Devstats, closed in GH: 56358, 57932, 63625, 64831, 64837, 64883, 65601
Open in Devstats, not in GH: 62723

So it seems tah we're missig close events quite often and probably one milestone change ... investigating each case step-by-step.

**Comment by lukaszgryglicki** [comment]
For example you mentioned 62692 (I've analyzed this first because you listed it as wrong while I didn't detected it as a wrong):
https://github.com/kubernetes/kubernetes/pull/62692
Its state was wrong yesterday, but is OK today because:
- 18 days ago `k8s-bot` removed the milestone (but due to how it works bot commented first (that was registered by devstats) and then removed milestone - which generates no GH event, so was not registered), 22 days ago was before my GH API fix.
-  22 hours ago bot commented "[APPROVALNOTIFIER] This PR is NOT APPROVED" and that generated GH events - PR state was autohealed.
- If the bot comment removing milestone happened say 10 days ago devstats would register:
  - GH comment event (bot's comment) which would not change the milestone
  - GH API milestone event that will remove the milestone.
Will see on the other PRs...

**Comment by lukaszgryglicki** [comment]
Open in Devstats, not listed in GH: 62723
Devstats state is: open, v.12
GH state is: no milestone, open (so it is not found in GH query then)
Bot removed v.12 milestone 19 days ago which is about the time when my GH API fix was applied...
Analysed events and I don't see any GH API events on this PR, maybe it was too early will check another ones.

**Comment by lukaszgryglicki** [comment]
Open in Devstats, closed in GH: 56358
Devstats state is: open, v.12
GH state is: merged, v.12
Bot merged 22 days ago - which generates no GH event.

**Comment by lukaszgryglicki** [comment]
57932: merged 21 days ago
63625: merged 22 days ago
64831: merged 21 days ago
64837: merged 22 days ago
64883: merged 21 days ago
65601: merged 12 days ago, and I see GH API events but no merge event, need to investigate this (only this PR has any GH API events)

**Comment by lukaszgryglicki** [comment]
Ohhhh it seems like for all those PRs correct events (merge, close) are registered (after GH API fix is started) but on the `gha_pull_requests` table, `gha_issues` only holds data shared between issue & PR - this is a GitHub data structure. 
Reinvestigating that way...

**Comment by lukaszgryglicki** [comment]
So now:
If You define materialized view for `PRs` not underlying issues like this:
```
create materialized view current_state.prs as
WITH pr_latest AS (
          SELECT prs.id,
             prs.dup_repo_id AS repo_id,
             prs.dup_repo_name AS repo_name,
             prs.number,
             prs.milestone_id,
             milestones.milestone,                                                                                   
             prs.state,
             prs.title,
             prs.user_id AS creator_id,
             prs.assignee_id,
             prs.created_at AS created_at,
             prs.updated_at,
             prs.closed_at,
             prs.merged_at,
             prs.body,
             prs.comments,
             row_number() OVER (PARTITION BY prs.id ORDER BY prs.updated_at DESC, prs.event_id DESC) AS rank
            FROM (gha_pull_requests prs
              LEFT JOIN current_state.milestones ON prs.milestone_id = milestones.id)
         )                                                                                                           
SELECT pr_latest.id,
     pr_latest.repo_id,
     pr_latest.repo_name,
     pr_latest.number,
     pr_latest.milestone_id,
     pr_latest.milestone,
     pr_latest.state,
     pr_latest.title,
     pr_latest.creator_id,
     pr_latest.assignee_id,
     pr_latest.created_at,
     pr_latest.updated_at,
     pr_latest.closed_at,
     pr_latest.merged_at,
     pr_latest.body,
     pr_latest.comments
    FROM pr_latest
   WHERE (pr_latest.rank = 1) 
   ORDER BY pr_latest.repo_name, pr_latest.number;
```

And then run: `select state, count(*) from current_state.prs where milestone='v1.12' group by state`

You will get "Open 7 Closed 16"
GitHub says "Open 6 Closed 16"

So we're off - investingating this case...

**Comment by lukaszgryglicki** [comment]
`select string_agg(number::text, ', ') from current_state.prs where milestone='v1.12' and state = 'open';`

The last problematic PR (65043) have milestone removed multiple times (and added) and it seems like we're missing last milestone removal ...

**Comment by lukaszgryglicki** [comment]
maybe it was due to API limit - but rather not.
Anyway - ~I think we will need an additional tool (not a part of the cron) that can run on a selected set of issues/PRs: by numbers, labels, milestones etc to query GH API on them manually and update state - will implement this asap.

**Comment by lukaszgryglicki** [comment]
@jberkus You are using `issues.dup_created_at AS created_at,` in you mat views definition.
`dup_created_at` comes from event creation date, so for example `issues` will have `created_at` from last event (rank = 1)
You should use:
```
issues.created_at AS created_at,
```
in all your mat views.

So its definition should be:
```
create materialized view current_state.issues as
WITH issue_latest AS (
          SELECT issues.id,
             issues.dup_repo_id AS repo_id,
             issues.dup_repo_name AS repo_name,
             issues.number,
             issues.is_pull_request,
             issues.milestone_id,
             milestones.milestone,
             issues.state,
             issues.title,
             issues.user_id AS creator_id,
             issues.assignee_id,
             issues.created_at AS created_at,
             issues.updated_at,
             issues.closed_at,
             issues.body,
             issues.comments,
             row_number() OVER (PARTITION BY issues.id ORDER BY issues.updated_at DESC, issues.event_id DESC) AS rank
            FROM (gha_issues issues
              LEFT JOIN current_state.milestones ON issues.milestone_id = milestones.id)
         )
SELECT issue_latest.id,
     issue_latest.repo_id,
     issue_latest.repo_name,
     issue_latest.number,
     issue_latest.is_pull_request,
     issue_latest.milestone_id,
     issue_latest.milestone,
     issue_latest.state,
     issue_latest.title,
     issue_latest.creator_id,
     issue_latest.assignee_id,
     issue_latest.created_at,
     issue_latest.updated_at,
     issue_latest.closed_at,
     issue_latest.body,
     issue_latest.comments
    FROM issue_latest
   WHERE (issue_latest.rank = 1)
   ORDER BY issue_latest.repo_name, issue_latest.number;
```

**Comment by lukaszgryglicki** [comment]
For issues case:
GH says: 17 open, 1 closed.
Devstats says: 16 open, 1 closed.
`select string_agg(number::text, ', ') from current_state.issues where milestone='v1.12' and state = 'open' and is_pull_request = false and repo_name = 'kubernetes/kubernetes'`

Devstats returns: 17130, 60712, 51540, 57618, 64164, 62747, 64627, 64530, 65122, 65021, 65392, 65225, 65266, 65718, 65494, 65972
GH returns: 65972, 65718, 65494, 65392, 65266, 65225, 65122, 65021, 64627, 64530, 64164, 60712, 58222, 58010, 57618, 51540, 17130

Diff is:
Devstats returns: 62747
GH returns: 58222, 58010

**Comment by lukaszgryglicki** [comment]
62747: we are missing final bot milestone removal, no idea why :-(
58222: bot added milestone on Jun 5 (so GH API fix was not there yet) but then bot commented 22 days ago - we don't have that event at all (seems like GHA missed this which is a known bug) no more events happened later, so I think this case is explained.
58010: we are missing last bots comment (from 12 days ago) but we're also missing luxas actions from 12 days ago which should be captured by GH API, no idea why currently

**Comment by lukaszgryglicki** [comment]
Summing up:
Some issues are from missing GHA events and then no GH API activity after 19 days ago.
But it seems like we're also missing some GH API activity sometimes, but in other cases all is OK.
@jberkus 
- Please use suggested `current_state.prs` for querying PRs related informations
- Please do not use `dup_created_at` as a issue/milestone creation date in your mat views because it refers to event creation date not milestopne/issue

I will:
- Double check my GH API code again now taking into account detected problems.
- I'll probably implement manual tool that will allow to force update of a given issues/PRs using GH API. its input will probably be a SQL query that should return list of issues/PRs to check.

**Comment by lukaszgryglicki** [comment]
So I've examined my GH API code and I think it should be OK. I was skipping some certain APi events (`referenced, mentioned, subscribed, unsubscribed, head_ref_deleted, head_ref_restored`) because they're not changing issue/PR state.
But maybe it is worth to record them in case we missed something before, so there is a more chances that we can get updated state.
I've did that.

**I still don't know why we missed few bot's merge API events yet :( **

I'll start debugging now and writing new tool for manual sync in the meantime - it will help debugging and I'll reuse 80% of code that I already have. I'll then run this tool with all debuging options enabled on our wrong cases to see what will happen.

unfortunatelly there are **no** chances to finish such a complex tool today - I'll start and will continue tomorrow.
@jberkus @dankohn

**Comment by lukaszgryglicki** [comment]
New manual sync tool ready, only need some debug info tweak in both manual and automatic tools.
kind of work in "dry run" mode without DB changing but pointing to what should be fixed.

Tomorrow I will have to tweak both tools for special debug info and see what events automatic tool receives (I hope there is a bug in *my code* and gitHub correctly sends all events) and what is the correct state reported by manual tools and when they differ and why.

That might be one full day of debugging becaus ethis stuff is a bit complex.

**Comment by jberkus** [comment]
Ok, let me know when you want me to check the data again.

**Comment by lukaszgryglicki** [comment]
Will do.
The best possible option (unrealistinc) is tomorrow - assuming that very deep debug will find some problem on my side
The other estimate is unknown.

But I will do my best to fix this ASAP.

**Comment by lukaszgryglicki** [comment]
back on this.

**Comment by lukaszgryglicki** [comment]
Hmm I've added a lot of debugging and I'm beginning to think that some events are *NOT* send by GH.
example:
- I've subscribed to k/k 62747
- I've waited 10 mins (to be sure)
- I've received *NO* GH event (issue events API) about my subscription

GH API says:
```
//     subscribed, unsubscribed
//       The Actor subscribed to or unsubscribed from notifications for an issue.
```

The good part is that I've added handlers to all possible current event types and even warning when new unknown/unhandled event is received. The more events - the smaller chance to miss any state change.

There are at least 5 event types not described in docs - that I'm catching:
`vendor//github.com/google/go-github/github/issues_events.go`:
```
  // Non specified in GH API but happenning
  eventTypes["review_requested"] = struct{}{}
  eventTypes["review_dismissed"] = struct{}{}
  eventTypes["review_request_removed"] = struct{}{}
  eventTypes["added_to_project"] = struct{}{}
  eventTypes["moved_columns_in_project"] = struct{}{}
```

Now I have a quite good debug/info infrastructure so I'm digging deeper.

**Comment by lukaszgryglicki** [comment]
I've subscribed and then (after some time) unsubscribed to all of PRs/Issues with incorrect state for k/k and milestone v1.12.
No GH API events were returned.
But the code (tripple/even quadrupple checked) seems to be ok and many other such "(un)subscribe" events are recorded... the good news is that manual sync tool is able to fix any arbitrary (given by sql) set of issues/PRs.
Also when processing all possible events I see no out-of-sync now... but I'm continuying digging.
Now I'll try to find "k8s-bot" merge event (recent from last few hours) that is not returned by GH API - if I find one then I'm quite sure GH API is not returning all possible repo events :/

**Comment by lukaszgryglicki** [comment]
I see an easy dirty hack solution:
- create cron job for manual sync tool with a specific select that will always keep some interesting group of issues/PRs correct.

This will make API usage higher and is ugly & dirty like hell - so I'm not attempting this atm. I still belive/hope something is wrong with my code.

**Comment by lukaszgryglicki** [comment]
I'm able to fix data (on the test server) - fixed all last weeks events. But I need to confirm all is OK with this, and also enchanced version of automatic sync will run during the weekend (on the test server only) - I hope that this enchanced version will not generate any out-of-sync data so manual hack won't be needed.

**Comment by jberkus** [comment]
Can you do a technical writeup of this that we can forward to GitHub?

**Comment by lukaszgryglicki** [comment]
Not yet. I don't want to write to Github when I'm not sure if the problem is not on my side.

**Comment by lukaszgryglicki** [comment]
We can eventually send a question to GH:

Hi, I'm using `ListRepositoryEvents`: `https://developer.github.com/v3/issues/events/#list-events-for-a-repository` to get a list of events for a selected repository.
Is it possible that this function does not return all events?

Documentation says:
```
  //     subscribed, unsubscribed
  //       The Actor subscribed to or unsubscribed from notifications for an issue.
```

I'm opening given repos Issue or PR (PR is also an Issue, right?), subscribing to it then eventually unsubscibing after 10 minutes and I'm not geting this event.
The bigger problem is that I'm not getting merge events when running this on `kubernetes/kubernetes` for some PRs merged by `k8s-merge-bot` - sometimes I receive merge event but sometimes not.

Best Regards.

cc @jberkus

**Comment by jberkus** [comment]
Can we isolate a specifc merge event we don't get?

**Comment by lukaszgryglicki** [comment]
For example this one: https://github.com/kubernetes/kubernetes/pull/65601
we are missing final bot merge event.
but to really check if no such event was fired I would have to call `ListRepositoryEvents` so many times on `k/k` repos to dig 18 days ago.
It returns 100 events max per page, and in a business days this is something about 1 hour.
This is why I've tested on (un)subscribe events myself.
I'll try to make a code that actually *will* dig so deep to be sure that we don't get this event

**Comment by lukaszgryglicki** [comment]
No... this is not this event.
We just had 3 events at the same exact second.
By my algorithm (even the first 20 day old version) deals with this and registers the final state.
And in case of this PR the final state *was* recorded OK - solution was the "current_state.prs" table instead of "current_state.issues".
For this https://github.com/kubernetes/kubernetes/issues/62747 I don't get the "demilestoned" event.

Now I'm in the middle of fixing data and it shouyld be OK by tomorrow, but I would ask GitHub the original question with (un)subscribing and no events.

**Comment by lukaszgryglicki** [comment]
I've fixed that data (manually updated all issues/PRs from v1.10, v1.11, v1.12 and with no milestone updated during last month) on both test & prod:
PRs:
```
gha=# select state, count(*) from current_state.prs where milestone='v1.12' and repo_name = 'kubernetes/kubernetes' group by state;
 state  | count 
--------+-------
 closed |    16
 open   |     7
(2 rows)
gha=# 
```
![screen shot 2018-07-16 at 12 10 55](https://user-images.githubusercontent.com/2469783/42753466-58b00588-88f1-11e8-8831-96c951d5b56e.png)

Issues:
```
gha=# select state, count(*) from current_state.issues where is_pull_request = false and milestone='v1.12' and repo_name = 'kubernetes/kubernetes' group by state;
 state  | count 
--------+-------
 closed |     2
 open   |    17
(2 rows)
```
![screen shot 2018-07-16 at 12 16 08](https://user-images.githubusercontent.com/2469783/42753717-1193f320-88f2-11e8-9c1a-4bc57b6b6425.png)

Numbers are now exactly the same.
I'll finish my started cncf/gitdm work (regenerate affiliations) and then I'll regenerate TSDB (dashboards) data on both test & prod.

https://k8s.devstats.cncf.io/d/22/open-issues-prs-by-milestone-and-repository?orgId=1&var-sig_name=All&var-milestone_name=v1.12&var-repo_name=kubernetes%2Fkubernetes&from=now-30d&to=now

**Comment by lukaszgryglicki** [comment]
I think data should be OK now. I was checking current state issues/PRs every day this week, and they were in sync with GitHub.

**Comment by lukaszgryglicki** [comment]
10 days in the row - number of open/closed issues/PRs in sync with GitHub.

**Comment by jberkus** [comment]
OK, seems like a pass.  Thanks so much for figuring out how to fix this!
